### PR TITLE
[DPO] add SLiC hinge loss to DPOTrainer

### PR DESCRIPTION
**Pull request by kashif**
from the [RSO paper]( https://arxiv.org/pdf/2309.06657.pdf) equation (12)

#816

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for adding this @kashif !

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -223,6 +226,7 @@ def __init__(
         self.padding_value = padding_value
 
         self.beta = beta
+        self.loss_type = loss_type
```
Comment (lvwerra) [inline trl/trainer/dpo_trainer.py:229]: I think we should raise an error if someone provided the wrong loss string. I guess currently they would get a "Losses referenced before definition" kind of error.

Otherwise looks good to me!

Comment (kashif) [reply trl/trainer/dpo_trainer.py:229]: ok fixing

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -364,6 +364,8 @@ def dpo_loss(
             losses = -F.logsigmoid(self.beta * logits)
         elif self.loss_type == "hinge":
             losses = torch.relu(1 - self.beta * logits)
+         else:
+            raise ValueError(f"Unknown loss type: {self.loss_type}")
```
Comment (lvwerra) [inline trl/trainer/dpo_trainer.py:368]: Small nit to make it more user-friendly :) 

```suggestion
            raise ValueError(f"Unknown loss type: {self.loss_type}. Should be one of ['sigmoid', 'hinge']")
```

Comment (lvwerra) [reply trl/trainer/dpo_trainer.py:368]: btw i think the indent is off
